hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/ByteBufferHeapData.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.serialization.impl;
+
+import com.hazelcast.nio.serialization.Data;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * A {@link Data} implementation where the content lives on the heap, and part of a byteBuffer
+ */
+public final class ByteBufferHeapData extends com.hazelcast.internal.serialization.impl.BaseHeapData {
+
+    private final ByteBuffer payload;
+    private final int length;
+
+    public ByteBufferHeapData(byte[] bytes, int offset, int length) {
+        this.length = length;
+        this.payload = ByteBuffer.wrap(bytes, offset, length).slice();
```

Comment:
Actually I am exchanging creating a large buffer with reusing it with offset. 
ByteBuffer reuses internal byte array. 
After reading your comment here https://github.com/hazelcast/hazelcast/pull/10629#pullrequestreview-40722488
How about making our HeapData work with offset instead of using ByteBuffer ?